### PR TITLE
rgw_sal_motr: [EOS-27520] Changes for empty bucket deletion.

### PR DESCRIPTION
**Pull request by SeagateChaDeepak**
# Change Summary

## Parent Story
_[EOS-27520](https://jts.seagate.com/browse/EOS-27520):_
_Delete bucket API does not work._

## Sub Tasks covered in this PR
_[EOS-28444](https://jts.seagate.com/browse/EOS-28444):_
_Delete empty bucket._

_[EOS-28445](https://jts.seagate.com/browse/EOS-28445):_
_Delete Bucket with Objects._

_[EOS-28446](https://jts.seagate.com/browse/EOS-28446):_
_Bucket stat and information update._

_[EOS-28447](https://jts.seagate.com/browse/EOS-28447):_
_Passing Ceph and Splunk tests for delete bucket._

## Exclusions in Implementation :
1. Abort multipart function is plugged in remove bucket, but internals to be implemented in a seperate ticket. _[EOS-27630](https://jts.seagate.com/browse/EOS-27630)_
2. Sync user stat function is plugged in remove bucket, but internals need to be implemented in a seperate ticket.
3. Forward to master function is plugged in remove bucket, but internals need to be implemented in a seperate ticket.
4. Notification for buckets currently is not present in motr sal. Need seperate ticket to address that.

## Implementation/Current Flow :
1. Removal of objects if delete children is True.
2. Plugged abort multipart function in place.
3. Deletion of multipart index.
4. Plugged sync user stats function in place.
5. Removing bucket information from user table.
6. Remove bucket info from bucket index.
7. Remove bucket instance information.
8. Plugged forward request to master function in place.

## Testing Done : 

<details>
  <summary> **Case 1 : Bucket Not Empty** </summary>

> [root@ssc-vm-g4-rhev4-0344 deepak_test_tools]# s3cmd --no-ssl put 1mfile s3://testbucket3
> upload: '1mfile' -> 's3://testbucket3/1mfile'  [1 of 1]
>  1048576 of 1048576   100% in    0s    12.33 MB/s  done
> 
> [root@ssc-vm-g4-rhev4-0344 deepak_test_tools]# s3cmd --no-ssl rb s3://testbucket3
> ERROR: S3 error: 409 (BucketNotEmpty)

</details>

<details>
  <summary> **Case 2 : Empty Bucket** </summary>

> [root@ssc-vm-g4-rhev4-0344 deepak_test_tools]# s3cmd --no-ssl rb s3://testbucket3
> Bucket 's3://testbucket3/' removed

</details>

<details>
<summary> **Case 3 : Reuse Bucket Name after deletion** </summary>

> [root@ssc-vm-g4-rhev4-0344 build]# s3cmd --no-ssl rb s3://dbucket
> Bucket 's3://dbucket/' removed
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 ls --endpoint-url http://localhost:8000
> 2022-02-04 00:46:47 mybucket
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 mb s3://dbucket --endpoint-url http://localhost:8000
> make_bucket: dbucket
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 ls --endpoint-url http://localhost:8000
> 2022-02-04 00:52:07 dbucket
> 2022-02-04 00:46:47 mybucket

</details>

<details>
<summary> **Case 4 : Use --recursive/--force switch for bucket deletion.** </summary>

> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 ls s3://mybucket --endpoint-url http://localhost:8000
> 2022-02-04 00:49:02    1048576 1mfile
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 rb s3://dbucket --endpoint-url http://localhost:8000
> remove_bucket failed: s3://dbucket An error occurred (BucketNotEmpty) when calling the DeleteBucket operation: Unknown
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 rb s3://dbucket --force --endpoint-url http://localhost:8000
> delete: s3://dbucket/1mfile
> remove_bucket: dbucket

</details>

<details>
<summary> Ceph and Splunk Tests passing for delete bucket </summary>

> (virtualenv) [root@ssc-vm-g4-rhev4-0344 s3-tests]# S3TEST_CONF=./s3tests.conf ./virtualenv/bin/nosetests s3tests_boto3.functional.test_s3:test_bucket_delete_nonempty -vv
> nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
> s3tests_boto3.functional.test_s3.test_bucket_delete_nonempty ... ok
> 
> ----------------------------------------------------------------------
> Ran 1 test in 0.517s
> 
> OK
> (virtualenv) [root@ssc-vm-g4-rhev4-0344 s3-tests]# S3TEST_CONF=./s3tests.conf ./virtualenv/bin/nosetests s3tests_boto3.functional.test_s3:test_bucket_create_delete -vv 
> nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
> s3tests_boto3.functional.test_s3.test_bucket_create_delete ... ok
> 
> ----------------------------------------------------------------------
> Ran 1 test in 0.369s
> 
> OK
> 
> (virtualenv) [root@ssc-vm-g4-rhev4-0344 s3-tests]# S3TEST_CONF=./s3tests.conf ./virtualenv/bin/nosetests s3tests_boto3.functional.test_s3:test_bucket_delete_notexist -vv
> nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
> s3tests_boto3.functional.test_s3.test_bucket_delete_notexist ... ok

</details>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

-----
[View rendered README.md](https://github.com/SeagateChaDeepak/cortx-rgw/blob/br/deepak/EOS-27520/README.md)

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -452,8 +452,67 @@ int MotrBucket::remove_bucket(const DoutPrefixProvider *dpp, bool delete_childre
...
+
+  ListParams params;
+  params.list_versions = true;
+  params.allow_unordered = true;
+
+  ListResults results;
+
+  results.objs.clear();
+
+  // TBD : support deletion of bucket with objects.
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:481]: This flag should be checked if we find bucket is not empty .. so should be shifted inside below if block
if (!results.objs.empty())

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -452,8 +452,67 @@ int MotrBucket::remove_bucket(const DoutPrefixProvider *dpp, bool delete_childre
...
+  if(delete_children){
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:482]: This flag should be checked if we find bucket is not empty .. so should be shifted inside below if block
if (!results.objs.empty())

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -3261,6 +3320,8 @@ int MotrStore::delete_motr_idx_by_name(string iname)
   struct m0_uint128 idx_id;
   struct m0_op *op = nullptr;
 
+  ldout(cctx, 0) << "delete_motr_idx_by_name=" << iname << dendl;
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:3323]: increase the debug level to 20 or more

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -3269,6 +3330,9 @@ int MotrStore::delete_motr_idx_by_name(string iname)
     goto out;
 
   m0_op_launch(&op, 1);
+
+  ldout(cctx, 0) << "waiting for op completion" << dendl;
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:3334]: increase the debug level to 70 or more

File: src/rgw/rgw_sal_motr.cc [hunk 5]
```diff
@@ -3279,6 +3343,8 @@ int MotrStore::delete_motr_idx_by_name(string iname)
   else if (rc < 0)
     ldout(cctx, 0) << "ERROR: index create failed: " << rc << dendl;
 
+  ldout(cctx, 0) << "delete_motr_idx_by_name rc=" << rc << dendl;
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:3346]: increase the debug level to 20 or more

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
work on comments

File: src/rgw/rgw_sal_motr.cc (hunk 5 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Done

File: src/rgw/rgw_sal_motr.cc (hunk 4 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Done

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Done

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Done

File: src/rgw/rgw_sal_motr.cc [hunk 6]
```diff
@@ -452,8 +452,123 @@ int MotrBucket::remove_bucket(const DoutPrefixProvider *dpp, bool delete_childre
...
+  } while(results.is_truncated);
+
+  // 2. Abort Mp uploads on the bucket.
+  ret = abort_multiparts(dpp, store->ctx());
+  if (ret < 0) {
+    return ret;
+  }
+
+  // 3. Remove mp index??
+  string bucket_multipart_iname = "motr.rgw.bucket." + info.bucket.name + ".multiparts";
```
Comment (bkirunge7) [inline src/rgw/rgw_sal_motr.cc:506]: please validate this scenario for expected behavior if bucket is empty and there are some ongoing multi parts, and delete_children is not true

**Review by kupranay — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 6 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc:506]: There is an abort_multipart function above this function call that will take care of this scenario, but its currently not implemented so cant test this scenario. JIRA for Abort Implementation (https://jts.seagate.com/browse/EOS-27630)

**Review by bkirunge7 — APPROVED** [review]

**Review by DPG17 — APPROVED** [review]

**Review by sachinpunadikar — APPROVED** [review]
+1

**Pull request merged**